charms/reactive/altrelations.py

```diff
@@ -41,9 +49,574 @@ def from_name(cls, relation_name):
+
+    Four flags are automatically managed for each endpoint. Endpoint handlers
+    can react to these flags using the :class:`~charms.reactive.decorators`.
+
+      * ``endpoint.{relation_name}.joined`` When the endpoint is :meth:`joined`
+      * ``endpoint.{relation_name}.changed``  When any relation data has changed
+      * ``endpoint.{relation_name}.changed.{field}``  When a specific field has changed
+      * ``endpoint.{relation_name}.departed`` When a remote unit is leaving
+
+    The ``joined`` flag will be automatically removed if all remote units leave
+    all relations, but the others must be manually removed by the interface
+    layer.
+
+    *Note that these flags can be used in the decorators of all handlers, not
+    just endpoint handlers, although this should be done with caution.*
+
+    The reactive framework automatically creates endpoint objects in the
+    :data:`~charms.reactive.helpers.context` namespace. Handlers can access the
+    endpoint objects from there, even if that endpoint doesn't have a joined
+    relation. The name of the endpoint object is based on the endpoints
+    configured in `metadata.yaml`. Note that endpoints will only be created
+    **if the interface layer of that endpoint is included in the `layer.yaml`
+    file.**
+
+    Endpoint handlers can iterate over the list of joined relations for an
+    endpoint via the :meth:`~charms.reactive.altrelations.Endpoint.relations`
+    collection.
+    """
+
+    @classmethod
+    def from_name(cls, relation_name):
+        return getattr(context.endpoints, relation_name, None)
+
+    @classmethod
+    def from_flag(cls, flag):
+        if '.' not in flag:
+            return None
+        parts = flag.split('.')
+        if parts[0] == 'endpoint':
+            return cls.from_name(parts[1])
+        else:
+            # some older handlers might not use the 'endpoint' prefix
+            return cls.from_name(parts[0])
+
+    @classmethod
+    def _startup(cls):
+        """
+        Populate context and manage automatic relation flags.
+        """
+        for relation_name in sorted(hookenv.relation_types()):
+            # populate context based on attached relations
+            relf = relation_factory(relation_name)
+            if not issubclass(relf, cls):
+                continue
+
+            rids = sorted(hookenv.relation_ids(relation_name))
+            endpoint = relf(relation_name, rids)
+            setattr(context.endpoints, relation_name, endpoint)
+            endpoint._manage_flags()
+            for relation in endpoint.relations:
+                hookenv.atexit(relation._flush_data)
+
+    def __init__(self, relation_name, relation_ids=None):
+        self._relation_name = relation_name
+        self._relations = KeyList(map(Relation, relation_ids or []),
+                                  key='relation_id')
+        self._all_units = None
+
+    @property
+    def relation_name(self):
```

Comment:
My thinking here (and for the `Endpoint` vs `EndpointInterface` above) was that the class represented the endpoint interface but specific instances represent actual relations (since they have an associated ID).  But I'm also feeling like that distinction might be too esoteric, and really the name is associated with the endpoint itself, not a specific relation.  So +1 on this and the above class name change.